### ISSUE TITLE
Insights on Chromecast latency control

### ISSUE THREAD

**Issue by ejurgensen** [philippe44/AirConnect#174]
I am the maintainer of [forked-daapd](https://github.com/ejurgensen/forked-daapd), which like AirConnect is able to stream to Chromecast (and is also C, btw). I am hoping to get in touch with you, because perhaps you have some insights on one of the issues I am trying to solve.

Until recently, forked-daapd worked like I think AirConnect does, by acting as a (mp3) streaming server, and getting the Chromecast to load and play a stream generated by forked-daapd. This had one major issue, which was that I couldn’t find a way of syncing the Chromecast device with other speakers that forked-daapd might be playing to, for instance Airplay speakers, or with the hosts soundcard. There was also another issue, which was that the Chromecasts were quite slow to start playback (some buffering issue, I imagine).

So I found out that Chromecasts actually support RTP via the so-called mirroring receiver app. That seemed interesting, since then sync might also be supported. So I changed forked-daapd to use RTP when casting. That did solve the “slow start” issue, but alas, I still can’t figure out how to control playback timing (except by simply delaying the RTP stream).

From your README I understand that you also have looked into latency control, and I understand that you haven’t found a way for Sonos. Is it the same situation for Chromecast? Have you by chance run into something that might help on this issue?

**Comment by philippe44** [comment]
oh, I did not know that CC has RTP support now, that's super interesting! I've made an application for the Logitech Media Server that synchronizes SB and AirPlay players, and as you know, AirPlay is using RTP. Is there some documentation on the mirroring receiver app? You need to setup some kind of shared reference using NTP and then reference RTP frames to it later, but that might be a bit platform specific

**Comment by ejurgensen** [comment]
I haven't found any actual documentation, I made my implementation by looking into Chromium's source code. You can see what I did here: https://github.com/ejurgensen/forked-daapd/blob/master/src/outputs/cast.c. Basically the session is setup with LAUNCH + OFFER and then you can check the calls in packets_send() to see how the RTP packets are constructed.

Yes, I did try to setup sync with NTP and RTCP, but my Chromecasts just ignored all that. So either it is unsupported, or I didn't do it the right way.

But what about the non-RTP method, i.e. the HTTP streaming method - have you found a way of controlling latency on Chromecast on that? If there is such a way I might scrap the RTP method and go back.

With regard to Airplay forked-daapd has its own implementation of RAOP - so no library. Check out https://github.com/ejurgensen/forked-daapd/blob/master/src/outputs/raop.c

**Comment by philippe44** [comment]
So you probably know as much or more than I do on this topic. Unfortunately, I've tried  to figure out the sync on the HTTP-based streaming and there is no doc, Google responded to me they will not make that public and I did not have the energy to try to reverse engineer it from the exchanges between devices (which might be encrypted anyway). I assume they have their own special way to extract the audio from the master of group that we would have to replicate, but no idea what kind of protocol they use.
